python/pyarrow/tests/test_serialization.py

```diff
@@ -0,0 +1,216 @@
+pa.lib.register_type(int, 20 * b"\x00", pickle=False,
+    custom_serializer=lambda obj: str(obj),
+    custom_deserializer=lambda serialized_obj: int(serialized_obj))
+if (sys.version_info < (3, 0)):
+    pa.lib.register_type(long, 20 * b"\x99", pickle=False,
+        custom_serializer=lambda obj: str(obj),
+        custom_deserializer=lambda serialized_obj: long(serialized_obj))
+
+if sys.version_info >= (3, 0):
+    long_extras = [0, np.array([["hi", u"hi"], [1.3, 1]])]
+else:
+    long_extras = [long(0), np.array([["hi", u"hi"], [1.3, long(1)]])]  # noqa: E501,F821
+
+PRIMITIVE_OBJECTS = [
+    0, 0.0, 0.9, 1 << 62, 1 << 100, 1 << 999,
+    [1 << 100, [1 << 100]], "a", string.printable, "\u262F",
+    u"hello world", u"\xff\xfe\x9c\x001\x000\x00", None, True,
+    False, [], (), {}, np.int8(3), np.int32(4), np.int64(5),
+    np.uint8(3), np.uint32(4), np.uint64(5), np.float32(1.9),
+    np.float64(1.9), np.zeros([100, 100]),
+    np.random.normal(size=[100, 100]), np.array(["hi", 3]),
+    np.array(["hi", 3], dtype=object)] + long_extras
+
+COMPLEX_OBJECTS = [
+    [[[[[[[[[[[[]]]]]]]]]]]],
+    {"obj{}".format(i): np.random.normal(size=[100, 100]) for i in range(10)},
+    # {(): {(): {(): {(): {(): {(): {(): {(): {(): {(): {
+    #       (): {(): {}}}}}}}}}}}}},
+    ((((((((((),),),),),),),),),),
+    {"a": {"b": {"c": {"d": {}}}}}]
+
+class Foo(object):
+    def __init__(self, value=0):
+        self.value = value
+
+    def __hash__(self):
+        return hash(self.value)
+
+    def __eq__(self, other):
+        return other.value == self.value
+
+
+class Bar(object):
+    def __init__(self):
+        for i, val in enumerate(PRIMITIVE_OBJECTS + COMPLEX_OBJECTS):
+            setattr(self, "field{}".format(i), val)
+
+
+class Baz(object):
+    def __init__(self):
+        self.foo = Foo()
+        self.bar = Bar()
+
+    def method(self, arg):
+        pass
+
+
+class Qux(object):
+    def __init__(self):
+        self.objs = [Foo(), Bar(), Baz()]
+
+
+class SubQux(Qux):
+    def __init__(self):
+        Qux.__init__(self)
+
+
+class CustomError(Exception):
+    pass
+
+
+Point = namedtuple("Point", ["x", "y"])
+NamedTupleExample = namedtuple("Example",
+                               "field1, field2, field3, field4, field5")
+
+
+CUSTOM_OBJECTS = [Exception("Test object."), CustomError(), Point(11, y=22),
+                  Foo(), Bar()] # ,  # Qux(), SubQux(),
+                  # NamedTupleExample(1, 1.0, "hi", np.zeros([3, 5]), [1, 2, 3])]
+
+pa.lib.register_type(Foo, 20 * b"\x02")
+pa.lib.register_type(Bar, 20 * b"\x03")
+pa.lib.register_type(Baz, 20 * b"\x04")
+pa.lib.register_type(Exception, 20 * b"\x05")
+pa.lib.register_type(CustomError, 20 * b"\x06")
+pa.lib.register_type(Point, 20 * b"\x07")
+pa.lib.register_type(NamedTupleExample, 20 * b"\x08")
+
+def serialization_roundtrip(value, f):
+    f.seek(0)
+    serialized, num_tensors = pa.lib.serialize_sequence(value)
+    pa.lib.write_python_object(serialized, num_tensors, f)
```

Comment:
Why not

`pa.serialize_sequence(value, f)`

and `result = pa.deserialize_sequence(f)` below?